### PR TITLE
[SU-176] Support leaving resources if enabled at the resource type level

### PR DESCRIPTION
**Pull request by MatthewBemis**
Please refer to https://docs.google.com/document/d/1y9s3EdmgiI9p_SAi_Z4mfp2ln5lPCZLlTAjQEHZyz4g/edit for technical design and motivation.

The TL;DR: Terra is very good at allowing users to share resources with other users, but extremely bad at allowing users to manage their own memberships. This takes the initial step of allowing users to leave a workspace that they no longer want access to. This has the user-facing benefit of allowing them to have more control and organization of their workspaces, and has the behind-the-scenes benefit of reducing Google Group membership counts. Related to that - in the past - if a user has reached the Google Group membership limit, they have been at the mercy of other Terra users who must be kind enough to remove them from a large number of managed groups, billing projects, and workspaces. This gives those power-users some autonomy to make their account useful again.

Ticket: <Link to Jira ticket>
<Put notes here to help reviewer understand this PR>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

**Comment by MatthewBemis** [comment]
jenkins retest

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -206,6 +206,7 @@ resourceTypes = {
       }
     }
     authDomainConstrainable = true
+    allowLeaving = true
```
Comment (MatthewBemis) [inline src/main/resources/reference.conf:209]: enabled only for `workspace`

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 2]
```diff
@@ -412,6 +412,31 @@ class ResourceService(
     eventualSubjects.toList.sequence.map(_.toMap)
   }
 
+  //Disallow orphaning, leaving when it's via a group, or leaving when the resource is public
+  def leaveResource(resourceType: ResourceType, resourceId: FullyQualifiedResourceId, samUser: SamUser, samRequestContext: SamRequestContext): IO[List[Boolean]] = {
+    accessPolicyDAO.listAccessPolicies(resourceId, samRequestContext) flatMap { policiesForResource =>
+      //1. Make sure that there are no public policies for this resource. Leaving a public resource is not supported.
```
Comment (nawatts) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:418]: Is this restriction due to some API limitation? If a public workspace has users with direct write/admin access, should they not be able to give up those permissions?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 3]
```diff
@@ -412,6 +412,31 @@ class ResourceService(
...
+
+      val policiesForUser = policiesForResource.filter(_.members.contains(samUser.id)).toList
+
+      //2. Make sure that the user has direct access that can be removed. Note that if user has both direct and indrect access,
+      //   this will succeed in removing their direct access. Which is still useful functionality.
+      if(policiesForUser.isEmpty) throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "You can only leave a resource that you have direct access to."))
+
+      //3. Make sure that removing the user will not orphan it, i.e. there must be at least one other owner on an "owning" policy
+      val ownerPolicies = policiesForUser.filter(_.roles.contains(resourceType.ownerRoleName))
+      val removalOrphansResource = ownerPolicies.exists(_.members.size <= 1)
```
Comment (nawatts) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:430]: I don't know how Sam policies are structured, but it seems surprising that this checks `policiesForUser` to determine if another user is an owner.

**Review by nawatts — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 2 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:418]: That's a fair point- this could be handled similarly to how we handle the case of removing direct membership when there is also indirect membership.

**Review by Shakespeared — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 4]
```diff
@@ -412,6 +412,31 @@ class ResourceService(
...
+      if(removalOrphansResource) throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "You may not leave a resource if you are the only owner. Please add another owner before leaving."))
+
+      //Any cases that we want to protect against are handled above, so it should be safe to proceed with removals
+      policiesForUser.parTraverse { policy =>
+        removeSubjectFromPolicy(policy.id, samUser.id, samRequestContext)
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:436]: I'm thinking we should have a log statement here

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 4 above)
Comment (MatthewBemis) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:436]: Added in fd52d311552e65329f301f7da7baaf4b896ce218

**Comment by MatthewBemis** [comment]
lotta test weirdness going on rn

jenkins retest

**Comment by MatthewBemis** [comment]
jenkins retest

**Comment by MatthewBemis** [comment]
jenkins retest

**Comment by MatthewBemis** [comment]
jenkins retest

**Comment by MatthewBemis** [comment]
The requested changes were handled in be2d9f7f58de226d3ca329b030e5b6168bc1f554. Greg is on vaca and from talking to him in person that was the only feedback he had so I am going to override the merge block and merge this since it has 2 other thumbs
